src/rgw/rgw_data_sync.cc

```diff
@@ -4136,12 +4138,78 @@ static bool has_olh_epoch(RGWModifyOp op) {
   return op == CLS_RGW_OP_LINK_OLH || op == CLS_RGW_OP_UNLINK_INSTANCE;
 }
 
+class RGWBucketShardIsDoneCR: public RGWCoroutine {
+  RGWDataSyncCtx *sc;
+  RGWDataSyncEnv *sync_env;
+  rgw_bucket_sync_status bucket_status;
+  const rgw_raw_obj status_obj;
+  const uint64_t num_shards;
+  const int shard_id;
+  RGWObjVersionTracker objv_tracker;
+  std::optional<next_bilog_result>& next_log;
+
+public:
+  RGWBucketShardIsDoneCR(RGWDataSyncCtx *_sc, const std::string& _status_oid,
+                         int _num_shards, int _shard_id, std::optional<next_bilog_result>& _next_log)
+    : RGWCoroutine(_sc->cct), sc(_sc), sync_env(_sc->env),
+      status_obj(sync_env->svc->zone->get_zone_params().log_pool, _status_oid),
+      num_shards(_num_shards), shard_id(_shard_id), next_log(_next_log) {}
+
+  int operate() override
+  {
+    reenter(this) {
+      do {
+      // read bucket sync status
+      using ReadCR = RGWSimpleRadosReadCR<rgw_bucket_sync_status>;
+      yield call(new ReadCR(sync_env->async_rados, sync_env->svc->sysobj,
+                            status_obj, &bucket_status, false, &objv_tracker));
+      if (retcode < 0) {
+        ldout(cct, 20) << "failed to read bucket shard status: "
+            << cpp_strerror(retcode) << dendl;
+        return set_cr_error(retcode);
+      }
+
+      yield {
+        // update bucket_status after a shard is done with current gen
+        bucket_status.shards_done_with_gen[shard_id] = true;
+
+        // increment gen if all shards are already done with current gen
+        const auto& done = bucket_status.shards_done_with_gen;
+        if (done.size() == num_shards) {
+          if (std::all_of(done.begin(), done.end(),
+            [] (const bool done){return done; } )) {
+            bucket_status.incremental_gen = next_log->generation;
+            bucket_status.shards_done_with_gen.clear();
+            bucket_status.shards_done_with_gen.resize(next_log->num_shards, false);
+          }
+        }
+        using WriteCR = RGWSimpleRadosWriteCR<rgw_bucket_sync_status>;
+        call(new WriteCR(sync_env->async_rados, sync_env->svc->sysobj,
+                          status_obj, bucket_status, &objv_tracker, false));
+        if (retcode < 0 && retcode != -ECANCELED) {
+          ldout(cct, 20) << "failed to write bucket sync status: " << cpp_strerror(retcode) << dendl;
+          return set_cr_error(retcode);
+        } else if (retcode >= 0) {
+          return set_cr_done();
+        }
+      }
+
+      } while (retcode == -ECANCELED);
```

Comment:
`objv_tracker` stores the version it read during `ReadCR`, and we passed that to the `WriteCR` so it would use `cls_version_check` to check whether the version changed since the read - that's where the ECANCELED error comes from

in this ECANCELED case, we're going to loop back and do the `ReadCR` again. but if we pass `objv_tracker` as it is, `ReadCR` would do that same check as `WriteCR`, and fail with ECANCELED too (since we know that version changed since our last read). so we need to call `objv_tracker.clear()` before we retry the read to skip that check in `RGWObjVersionTracker::prepare_op_for_read()`. this way, the next `ReadCR` just uses `cls_version_read` to read the updated version number for our next call to `WriteCR`

Reply:
oops! fixed it. thanks